### PR TITLE
The Photography Update - refactored photos, album persistence, large photos support and round based saving in logs

### LINKED ISSUE #7092: Photographs don't show janicart nor windows
Photographs do not show whole-tile windows.

They also do not capture the janicart.

They also do not get fire lights.

Nor Plasma.

Fix that shit.

### LINKED ISSUE #11276: Cyborg photo system
As a cyborg, the "view images" button always returns with "No Images saved", rather than the drop down of your pictures, no matter how many photos you have saved.
Printing works normally and gives you a list of the photos you took however.

Unsure if the same issue comes with the AI.

### LINKED ISSUE #23184: Taking photos with cameras is fucky
About 7 times out of 10, if you try to take a photograph of something with a camera, nothing will happen. I have no idea why or what's causing this, but it's on the server right now, so it isn't related to my picture frames PR.

### PR DESCRIPTION
**Pull request by silicons**
- [ ] photo album persistence based on id
- [x] save photos based on round in log folders

Adds support for custom sized photos
Refactor photos and photography to use the datum photo
Tweaks to silicon photo selection to make it useful and better
Cyborgs now sync photos with their ai on lawsync 
Photos are now stored every round for Atlanta Ned to put on his site so we can laugh at embarresing moments
Planned: captain photo album persistence along with others.

Fixes #7092
Fixes #23184
Fixes #11276

:cl: Skylar
rscadd: Cameras now support photos up to 15x15 in size, and independent x/y tile heights!
refactor: Photos and photography in general will use /datum/photo to store photo information, instead of /obj/item/weapon/picture.
tweak: Cyborgs can now print pictures for a cheaper cost.
tweak: Cyborgs now properly sync pictures with the AI and grabs pictures with the AI.
experimental: Photos are now logged by round ID with filenames of their numerical id and a json of their information if the variable is enabled!
MORE TO COME----

### CODE REVIEW

File: code/game/gamemodes/miniantags/abduction/machinery/experiment.dm [hunk 1]
```diff
@@ -98,9 +98,8 @@
 	var/dat
 	dat += "<h3> Experiment </h3>"
 	if(occupant)
-		var/obj/item/weapon/photo/P = new
-		P.photocreate(null, icon(dissection_icon(occupant), dir = SOUTH))
-		user << browse_rsc(P.img, "dissection_img")
+		var/icon/I = icon(dissection_icon(occupant), dir = SOUTH)
```
Comment (RemieRichards) [inline code/game/gamemodes/miniantags/abduction/machinery/experiment.dm]: lol @AnturK you hack.

**Comment by Tacolizard** [comment]
dat changelog

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 2]
```diff
@@ -1097,6 +1097,10 @@
 /mob/living/silicon/robot/shell
 	shell = TRUE
 
+/mob/living/silicon/robot/proc/picturesync()
+	if(connected_ai && connected_ai.aicamera && aicamera)
+		connected_ai.aicamera.stored |= aicamera.stored
```
Comment (RemieRichards) [inline code/modules/mob/living/silicon/robot/robot.dm]: never use ```list |= val``` where you can use ```list[val] = TRUE```

**Comment by Shadowlight213** [comment]
\>picture persistence.
No more persistence features. Just stop

**Comment by kevinz000** [comment]
@Shadowlight213 what's the worst that can happen with making photo albums have a use?

File: code/modules/photography/_photos.dm [hunk 3]
```diff
@@ -0,0 +1,36 @@
+
+/datum/picture
+	var/picture_name = "picture"
+	var/picture_desc = "This is a picture."
+	var/icon/picture_image
```
Comment (RemieRichards) [inline code/modules/photography/_photos.dm]: fat lie, this is an ```/icon```

File: code/modules/photography/_photos.dm [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	P.picture_image = picture_image
+	P.picture_icon = picture_icon
+	P.psize_x = psize_x - cropx * 2
+	P.psize_y = psize_y - cropy * 2
+	P.has_blueprints = has_blueprints
+	if(greyscale)
+		P.picture_image.MapColors(rgb(77,77,77), rgb(150,150,150), rgb(28,28,28), rgb(0,0,0))
+		P.picture_icon.MapColors(rgb(77,77,77), rgb(150,150,150), rgb(28,28,28), rgb(0,0,0))
+	if(cropx || cropy)
+		P.picture_image.Crop(cropx, cropy, psize_x - cropx, psize_y - cropy)
```
Comment (RemieRichards) [inline code/modules/photography/_photos.dm]: so in the above, you apply the operation to both icons.
but here only to one icon?

Why do you even need the second icon?
And why aren't you modifying it here? (is it some kind of immutable icon that represents the picture as it was taken?)

**Comment by nfreader** [comment]
>Photos are now stored every round for Atlanta Ned to put on his site so we can laugh at embarresing moments 

See if you can include captions as well :)

File: code/modules/photography/camera.dm [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	if(C)
+		pictures_max = C.pictures_max
+		pictures_left = C.pictures_left
+		visible_message("[C] has been imbued with godlike power!")
+		qdel(C)
+
+/obj/item/device/camera/spooky
+	name = "camera obscura"
+	desc = "A polaroid camera, some say it can see ghosts!"
+	see_ghosts = 1
```
Comment (RemieRichards) [inline code/modules/photography/camera.dm]: ```TRUE```

**Comment by Shadowlight213** [comment]
My complaint is filesize mainly

File: code/modules/photography/ai_camera.dm [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		to_chat(usr, "<span class='warning'>Insufficent toner to print image.</span>")
+		return
+	var/datum/picture/selection = selectpicture()
+	if(!istype(selection))
+		to_chat(usr, "<span class='warning'>Invalid Image.</span>")
+		return
+	var/obj/item/weapon/photo/p = new /obj/item/weapon/photo(C.loc, selection)
+	p.pixel_x = rand(-10, 10)
+	p.pixel_y = rand(-10, 10)
+	C.toner -= 20	 //Cyborgs are very ineffeicient at printing an image
```
Comment (GunHog) [inline code/modules/photography/ai_camera.dm]: Could you make this cheaper? I doubt more than a handful of people would object if a borg could print out lots of photos.

Comment (kevinz000) [inline code/modules/photography/ai_camera.dm]: Sure

File: code/controllers/subsystem/persistence.dm [hunk 7]
```diff
@@ -149,11 +149,17 @@ SUBSYSTEM_DEF(persistence)
...
 	CollectChiselMessages()
 	CollectSecretSatchels()
 	CollectTrophies()
+	StorePhotoInformation()
+
+/datum/controller/subsystem/persistence/proc/StorePhotoInformation()
+	var/finalpath = GLOB.picture_log_folder
+	finalpath += "photos.json"
+	var/file = file(finalpath)
+	file << json_encode(GLOB.picture_logging_information)
```
Comment (Cyberboss) [inline code/controllers/subsystem/persistence.dm]: Use text2file

File: code/modules/photography/_photos.dm [hunk 8]
```diff
@@ -1,21 +1,45 @@
 
+GLOBAL_VAR_INIT(picture_datum_id, 1)
+//GLOBAL_PROTECT(picture_datum_id)			//please don't.
+GLOBAL_LIST_INIT(picture_logging_information, list())
+//GLOBAL_PROTECT(picture_logging_information)			//pls no.
```
Comment (Cyberboss) [inline code/modules/photography/_photos.dm]: You're turning these on eventually right?

File: code/modules/photography/_photos.dm [hunk 9]
```diff
@@ -1,21 +1,45 @@
 
+GLOBAL_VAR_INIT(picture_datum_id, 1)
+//GLOBAL_PROTECT(picture_datum_id)			//please don't.
+GLOBAL_LIST_INIT(picture_logging_information, list())
```
Comment (Cyberboss) [inline code/modules/photography/_photos.dm]: GLOBAL_LIST_EMPTY(picture_logging_information)

File: code/modules/photography/camera.dm [hunk 10]
```diff
@@ -103,6 +111,7 @@
 			desc += camera_get_mobs(T)
 			if(locate(/obj/item/areaeditor/blueprints) in T)
 				blueprints = TRUE
+			CHECK_TICK
```
Comment (Cyberboss) [inline code/modules/photography/camera.dm]: This is just going to fuck up the photos

File: code/modules/photography/camera.dm [hunk 11]
```diff
@@ -136,6 +146,7 @@
 					holding += " and \a [I]"
 			holding = holding.Join()
 			mob_details += "You can also see [L] on the photo[L.health < (L.maxHealth * 0.75) ? " - [L] looks hurt":""].[holding ? " [holding]":"."]."
+		CHECK_TICK
```
Comment (Cyberboss) [inline code/modules/photography/camera.dm]: Ditto

File: code/modules/photography/camera.dm [hunk 12]
```diff
@@ -173,6 +184,7 @@
 			yo += AM.step_y
 		var/icon/img = getFlatIcon(A)
 		res.Blend(img, blendMode2iconMode(A.blend_mode), xo, yo)
+		CHECK_TICK
```
Comment (Cyberboss) [inline code/modules/photography/camera.dm]: Ditto

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Iatots** [comment]
What do you mean by ID? Your C-key or whatever name you select at roundstart? Can I see and potentially mess with other people's albums by using their names?

**Comment by kevinz000** [comment]
@Iatots every picture generated has an ID used internally.

File: code/modules/photography/_photos.dm (hunk 8 above)
Comment (kevinz000) [inline code/modules/photography/_photos.dm]: yep, it's disabled for testing, but leave the comment so i don't forget.

File: code/modules/photography/_photos.dm (hunk 4 above)
Comment (kevinz000) [inline code/modules/photography/_photos.dm]: fixed, second icon is for photos having a small inhand icon that approximates its contents.

File: code/modules/photography/camera.dm (hunk 10 above)
Comment (kevinz000) [inline code/modules/photography/camera.dm]: fixed, cameras now grab all atoms once.

File: code/modules/photography/camera.dm (hunk 11 above)
Comment (kevinz000) [inline code/modules/photography/camera.dm]: fixed, cameras now grab all atoms once.

File: code/modules/photography/camera.dm (hunk 12 above)
Comment (kevinz000) [inline code/modules/photography/camera.dm]: fixed, cameras now grab all atoms once.

**Comment by Tacolizard** [comment]
>🆑 Skylar

**Comment by Tacolizard** [comment]
you also didn't put a closing cl tag at the end of the changelog you weenie

**Comment by Cyberboss** [comment]
@kevinz000

File: code/modules/admin/verbs/mapping.dm [hunk 13]
```diff
@@ -281,3 +282,64 @@ GLOBAL_VAR_INIT(say_disabled, FALSE)
...
+				continue
+			atoms[T] = TRUE
+
+	var/list/sorted = list()
+	var/j
+	for(var/i = 1 to atoms.len)
+		var/atom/c = atoms[i]
+		for(j = sorted.len, j > 0, --j)
+			var/atom/c2 = sorted[j]
+			if(c2.layer <= c.layer)
```
Comment (RemieRichards) [inline code/modules/admin/verbs/mapping.dm]: >doesn't even handle planes

:/

File: code/modules/admin/verbs/mapping.dm [hunk 14]
```diff
@@ -281,3 +282,64 @@ GLOBAL_VAR_INIT(say_disabled, FALSE)
...
+
+#ifdef TESTING
+	var/input = input("This command takes a full picture of the current zlevel and saves it to the server directory. THIS WILL LOCK UP THE SERVER AND GAME FOR A MEDIUM TO LONG PERIOD OF TIME \
+	DEPENDING ON HOW FAST THE HOST IS. ARE YOU SURE YOU WISH TO CONTINUE?", "ABORT", "NO") as text|null in list("NO", "YES")
+	if(input == "NO")
+		return
+
+	full_zlevel_picture(client.mob.z)
+
+/proc/full_zlevel_picture(z)
```
Comment (RemieRichards) [inline code/modules/admin/verbs/mapping.dm]: HAHA wowee.
>getFlatIcon
>one of the slowest procs
>on all atoms on the Z level that are visible

No wonder it's locked behind debug stuff.

**Comment by kevinz000** [comment]
@Cyberboss I need your transit stuff since you insist on instant copying.

File: code/modules/admin/verbs/mapping.dm (hunk 13 above)
Comment (kevinz000) [inline code/modules/admin/verbs/mapping.dm]: ......??????????!!!

Comment (RemieRichards) [inline code/modules/admin/verbs/mapping.dm]: planes are like layers, they control the visual order of things.

**Comment by kevinz000** [comment]
im doing this when transit refactor is done

**Comment by nicbn** [comment]
Kevin do your shit

**Comment by nicbn** [comment]
Kevinz finish your shit

**Comment by Bawhoppen** [comment]
Finish this seriously, we need it

**Pull request closed without merging**